pkg/loglines/dedup.go

```diff
@@ -0,0 +1,217 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+package loglines
+
+import (
+	"fmt"
+	"os"
+	"sort"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	keepGenerations = 5
+)
+
+func dedupGeneration(lineSlices []*LineSlice) *LineSlice {
+	ls1, ls2 := DedupLineSlices(lineSlices)
+	output := MakeSliceFromLines(MakeNewCombinedSlice(ls1.Lines, ls2.Lines), ls2.Generation)
+	return output
+}
+
+// findLineIndex will find the index of a line in a slice of lines
+// and will return -1 if it is not found
+func findLineIndex(needle *ProcessedLine, list []*ProcessedLine) int {
+	checkLine := needle.Full
+	for i, hay := range list {
+		if hay.Full == checkLine {
+			return i
+		}
+	}
+	return -1
+}
+
+//nolint:varnamelen // x and y are just two sets of lines
+func findIncompatableLines(x, y []*ProcessedLine) int {
+	for index, line := range x {
+		if index >= len(y) {
+			// we've reached the end of y so no bad lines found
+			break
+		}
+		if line.Full != y[index].Full {
+			return index
+		}
+	}
+	return -1
+}
+
+func findNextMatching(a, b []*ProcessedLine) (offset, index int) {
+	index = findLineIndex(a[offset], b)
+	if index == -1 {
+		// Can't find X in Y, Search for next line which is in Y
+		for offset = 1; offset < len(a); offset++ {
+			index = findLineIndex(a[offset], b)
+			if index != -1 {
+				break
+			}
+		}
+	}
+	return offset, index
+}
+
+//nolint:gocritic,varnamelen // don't want to name the return values as they should be built later, I think x, y are expressive enough names
+func fixLines(x, y []*ProcessedLine, issueIndex int) ([]*ProcessedLine, []*ProcessedLine, bool) {
```

Comment:
fixLines needs a sentence or two docstring explaining what the two output line slices are and what guarantees are made about them